### PR TITLE
feat: make ak.mean and ak.count differentiable

### LINKED ISSUE #2638: Differentiating through an ak.mean
### Version of Awkward Array

main branch

### Description and code to reproduce

This is a follow-up to #2591 with a slightly more simplified setup. It should be conceptually possible to differentiate through taking a mean. Currently this does not work.

Reproducer:
```python
import awkward as ak
import jax
import uproot

ak.jax.register_and_check()

ttbar_file = "https://github.com/scikit-hep/scikit-hep-testdata/"\
    "raw/main/src/skhep_testdata/data/nanoAOD_2015_CMS_Open_Data_ttbar.root"

def mean_jet_pt(jets):
    return ak.mean(jets.pt)

with uproot.open(ttbar_file) as f:
    arr = f["Events"].arrays(["Jet_pt","Jet_eta", "Jet_phi", "Jet_mass"])
    evtfilter = ak.num(arr["Jet_pt"]) >= 2
    jets = ak.zip(dict(zip(["pt","eta", "phi", "mass"], ak.unzip(arr))), with_name="Momentum4D")[evtfilter]
    jets = ak.to_backend(jets, "jax")


jax.value_and_grad(mean_jet_pt, argnums=0)(jets)
```

Result:
```pytb
RuntimeError: Cannot differentiate through count_zero

This error occurred while calling

    ak.mean(
        <Array [[...], [...], ..., [...], [...]] type='140 * var * float32'>
    )
```

A standalone jax version of taking a mean works fine:
```python
import jax.numpy as jnp

def mean(j):
    return jnp.mean(j)
    
data = jnp.array([1, 7, 3, 5],dtype=float)

jax.value_and_grad(mean, argnums=0)(data)
```

### PR DESCRIPTION
**Pull request by Saransh-cpp**
Fixes #2638

Relevant discussion on slack - https://iris-hep.slack.com/archives/C0155BGPGE4/p1707239216196289

cc: @alexander-held

### CODE REVIEW

File: src/awkward/_connect/jax/reducers.py [hunk 1]
```diff
@@ -111,7 +111,16 @@ def apply(
         shifts: ak.index.Index | None,
         outlength: ShapeItem,
     ) -> ak.contents.NumpyArray:
-        raise RuntimeError("Cannot differentiate through count_zero")
+        assert isinstance(array, ak.contents.NumpyArray)
+        result = jax.numpy.ones_like(array.data, dtype=array.dtype)
+        result = jax.ops.segment_sum(result, parents.data)
+
+        if array.dtype.type in (np.complex128, np.complex64):
```
Comment (agoose77) [inline src/awkward/_connect/jax/reducers.py]: I think we would be better to check "is this complex"
```suggestion
        if np.issubdtype(array.dtype, np.complexfloating):
```

Comment (Saransh-cpp) [reply src/awkward/_connect/jax/reducers.py]: I see, thanks!

**Comment by agoose77** [comment]
@Saransh-cpp is this stuck on anything? Just checking whether I can offer any help.

**Comment by Saransh-cpp** [comment]
Hi @agoose77, the PR is ready for review; not stuck on anything!

File: src/awkward/_connect/jax/reducers.py [hunk 2]
```diff
@@ -169,7 +178,7 @@ def apply(
             return ak.contents.NumpyArray(
                 array.backend.nplike.asarray(result, dtype=array.dtype)
             )
-        elif array.dtype.type in (np.complex128, np.complex64):
+        elif np.issubdtype(array.dtype, np.complexfloating):
```
Comment (jpivarski) [inline src/awkward/_connect/jax/reducers.py:181]: Because JAX uses 32-bit floats in its complex numbers. Got it.

**Review by jpivarski — APPROVED** [review]
This looks good; I'll approve it!

And I'll merge it because it looks done.

**Pull request merged**